### PR TITLE
`u.sub(0)` should still be in a `ComponentFunctionSpace` if the vector is 1 dimensional

### PR DESCRIPTION
**Pull request by JHopeCollins**
Currrently, for `u` in a `VectorFunctionSpace`, the `Functions` returned by `u.sub` are in different types of spaces depending on the dimension of the VFS.

- if `dim` is 2 or higher, then `u.sub(i).function_space()` is a `ComponentProxyFunctionSpace`.
- if `dim` is 1, then `u.sub(0) is u`, instead of being in the `ComponentProxyFunctionSpace`.

This PR changes the `u.sub(0)` for 1-dim vectors to be in the `ComponentProxyFunctionSpace` by changing the if-statement to test `self.rank` instead of `self.dof_dset.cdim`.
Change made here: https://github.com/firedrakeproject/firedrake/blob/a2a944801ba1ed3404b90a707502285bbc94afa4/firedrake/function.py#L127

The rest of the changes are making how/where the mixed/vector-ness is tested in subfunctions/sub/_components consistent across functions and function spaces.

### CODE REVIEW

File: firedrake/function.py [hunk 1]
```diff
@@ -143,9 +147,12 @@ def sub(self, i):
...
         indexing the ith component of the space, suitable for use in
         boundary condition application."""
-        if len(self.function_space()) == 1:
-            return self._components[i]
-        return self.subfunctions[i]
+        mixed = len(self.function_space()) != 1
+        data = self.subfunctions if mixed else self._components
+        bound = len(data)
+        if i < 0 or i >= bound:
+            raise IndexError("Invalid component %d, not in [0, %d)" % (i, bound))
```
Comment (connorjward) [inline firedrake/function.py:154]:
```suggestion
            raise IndexError(f"Invalid component {i}, not in [0, {bound})")
```

File: firedrake/function.py [hunk 2]
```diff
@@ -124,12 +124,16 @@ def split(self):
...
-            return (self, )
+        if self.function_space().rank == 0:
+            return tuple((self, ))
         else:
-            return tuple(CoordinatelessFunction(self.function_space().sub(i), val=op2.DatView(self.dat, j),
-                                                name="view[%d](%s)" % (i, self.name()))
-                         for i, j in enumerate(np.ndindex(self.dof_dset.dim)))
+            if self.dof_dset.cdim == 1:
+                return tuple((CoordinatelessFunction(self.function_space().sub(0), val=self.dat,
+                                                     name="view[%d](%s)" % (0, self.name())),))
```
Comment (connorjward) [inline firedrake/function.py:132]:
```suggestion
                return (CoordinatelessFunction(self.function_space().sub(0), val=self.dat,
                                               name=f"view[0]({self.name()})"),)
```

File: firedrake/function.py [hunk 3]
```diff
@@ -124,12 +124,16 @@ def split(self):
...
+            else:
+                return tuple(CoordinatelessFunction(self.function_space().sub(i), val=op2.DatView(self.dat, j),
+                                                    name="view[%d](%s)" % (i, self.name()))
```
Comment (connorjward) [inline firedrake/function.py:135]:
```suggestion
                                                    name=f"view[{i}]({self.name()})")
```

File: firedrake/function.py [hunk 4]
```diff
@@ -345,9 +352,12 @@ def sub(self, i):
...
         indexing the ith component of the space, suitable for use in
         boundary condition application."""
-        if len(self.function_space()) == 1:
-            return self._components[i]
-        return self.subfunctions[i]
+        mixed = len(self.function_space()) != 1
+        data = self.subfunctions if mixed else self._components
+        bound = len(data)
+        if i < 0 or i >= bound:
+            raise IndexError("Invalid component %d, not in [0, %d)" % (i, bound))
```
Comment (connorjward) [inline firedrake/function.py:359]:
```suggestion
            raise IndexError(f"Invalid component {i}, not in [0, {bound})")
```

File: firedrake/functionspaceimpl.py [hunk 5]
```diff
@@ -182,10 +182,12 @@ def _components(self):
 
     @PETSc.Log.EventDecorator()
     def sub(self, i):
-        bound = len(self._components)
+        mixed = len(self) != 1
+        data = self.subfunctions if mixed else self._components
+        bound = len(data)
         if i < 0 or i >= bound:
             raise IndexError("Invalid component %d, not in [0, %d)" % (i, bound))
```
Comment (connorjward) [inline firedrake/functionspaceimpl.py:189]:
```suggestion
            raise IndexError(f"Invalid component {i}, not in [0, {bound})")
```

File: firedrake/functionspaceimpl.py [hunk 6]
```diff
@@ -639,7 +641,7 @@ def __str__(self):
     @utils.cached_property
     def subfunctions(self):
         r"""Split into a tuple of constituent spaces."""
-        return (self, )
+        return tuple((self, ))
```
Comment (connorjward) [inline firedrake/functionspaceimpl.py:644]:
```suggestion
        return (self, )
```
`tuple` not needed.

File: firedrake/functionspaceimpl.py [hunk 7]
```diff
@@ -654,13 +656,16 @@ def __getitem__(self, i):
...
+            return tuple(ComponentFunctionSpace(self, i) for i in range(self.block_size))
 
     def sub(self, i):
         r"""Return a view into the ith component."""
-        if self.rank == 0:
-            assert i == 0
-            return self
+        bound = len(self._components)
+        if i < 0 or i >= bound:
+            raise IndexError("Invalid component %d, not in [0, %d)" % (i, bound))
```
Comment (connorjward) [inline firedrake/functionspaceimpl.py:668]:
```suggestion
            raise IndexError(f"Invalid component {i}, not in [0, {bound})")
```

**Review by connorjward — CHANGES_REQUESTED** [review]
Basically just f-strings

**Review by connorjward — APPROVED** [review]

File: firedrake/function.py [hunk 8]
```diff
@@ -124,12 +124,16 @@ def split(self):
 
     @utils.cached_property
     def _components(self):
-        if self.dof_dset.cdim == 1:
-            return (self, )
+        if self.function_space().rank == 0:
+            return tuple((self, ))
```
Comment (pbrubeck) [inline firedrake/function.py:128]: `tuple` is not needed here

File: firedrake/function.py [hunk 9]
```diff
@@ -327,8 +334,8 @@ def split(self):
 
     @utils.cached_property
     def _components(self):
-        if self.function_space().block_size == 1:
-            return (self, )
+        if self.function_space().rank == 0:
+            return tuple((self, ))
```
Comment (pbrubeck) [inline firedrake/function.py:338]: Same here

**Pull request merged**